File: client/src/api/notifications.ts
import { API_URL } from "@/config";
import { getAuthHeaders } from "@/lib/utils";
import { parseApiError } from "@/lib/api-error";
export interface Notification {
    uid: string;
    user_username: string;
    actor_username: string;
    actor_avatar: string;
    actor_is_anonymous?: boolean;
    type: string;
    reference_uid: string;
    post_uid: string;
    content: string;
    question_content?: string;
    is_read: boolean;
    created_at: string;
}
export async function listNotifications(limit?: number, offset?: number): Promise<Notification[]> {
    const params = new URLSearchParams({
        ...(limit !== undefined ? { limit: limit.toString() } : {}),
        ...(offset !== undefined ? { offset: offset.toString() } : {}),
    });
    const res = await fetch(`${API_URL}/users/me/notifications?${params}`, {
        method: "GET",
        headers: {
            ...getAuthHeaders(),
        },
    });
    if (!res.ok) await parseApiError(res);
    return res.json();
}

export async function markNotificationsRead(): Promise<void> {
  await fetch(`${API_URL}/users/me/notifications/read`, {
    method: "POST",
    headers: { ...getAuthHeaders() },
  });
}

export async function getUnreadNotificationCount(): Promise<number> {
    const res = await fetch(`${API_URL}/users/me/notifications/unread-count`, {
        headers: { ...getAuthHeaders() },
    });
    if (!res.ok) await parseApiError(res);
    const { count } = await res.json();
    return count;
}

File: server/internal/repository/notifications.go
package repository

import (
	"context"
	"echo/internal/database"
	"echo/internal/types"
	"fmt"

	"github.com/google/uuid"
)

func (r *Repository) ListNotifications(ctx context.Context, currentUser string, limit, offset int32) ([]types.Notification, error) {
	rows, err := r.Q.ListNotifications(ctx, database.ListNotificationsParams{
		UserUsername: currentUser,
		Limit:        limit,
		Offset:       offset,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to fetch notifications: %w", err)
	}

		var notifications []types.Notification

		for _, row := range rows {

			n := types.Notification{

				UID:             uuid.UUID(row.Uid.Bytes).String(),

				UserUsername:    row.UserUsername,

				Type:            row.Type,

				ReferenceUID:    uuid.UUID(row.ReferenceUid.Bytes).String(),

				IsRead:          row.IsRead.Bool,

				CreatedAt:       row.CreatedAt.Time,

				Content:         row.Content,

				QuestionContent: row.QuestionContent,

			}

			if row.ActorAvatar.Valid {

				n.ActorAvatar = row.ActorAvatar.String

			}

			if row.ActorUsername.Valid {

				n.ActorUsername = row.ActorUsername.String

			}

			notifications = append(notifications, n)

		}

		return notifications, nil

	}

	